Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp

```diff
@@ -147,10 +209,76 @@ RefPtr<TypeDefinition> TypeDefinition::tryCreateStructType(StructFieldCount fiel
     void* memory = nullptr;
     if (!result.getValue(memory))
         return nullptr;
-    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(fieldCount);
+    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(TypeDefinitionKind::StructType, fieldCount);
+    return adoptRef(signature);
+}
+
+RefPtr<TypeDefinition> TypeDefinition::tryCreateRecursionGroup(RecursionGroupCount typeCount)
+{
+    // We use WTF_MAKE_FAST_ALLOCATED for this class.
+    auto result = tryFastMalloc(allocatedRecursionGroupSize(typeCount));
+    void* memory = nullptr;
+    if (!result.getValue(memory))
+        return nullptr;
+    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(TypeDefinitionKind::RecursionGroup, typeCount);
     return adoptRef(signature);
 }
 
+RefPtr<TypeDefinition> TypeDefinition::tryCreateProjection()
+{
+    // We use WTF_MAKE_FAST_ALLOCATED for this class.
+    auto result = tryFastMalloc(allocatedProjectionSize());
+    void* memory = nullptr;
+    if (!result.getValue(memory))
+        return nullptr;
+    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(TypeDefinitionKind::Projection);
+    return adoptRef(signature);
+}
+
+Type TypeDefinition::substitute(Type type, TypeIndex projectee)
+{
+    if (type.kind == TypeKind::Rec) {
+        RefPtr<TypeDefinition> projection = TypeInformation::typeDefinitionForProjection(projectee, static_cast<ProjectionIndex>(type.index));
+        TypeKind kind = type.isNullable() ? TypeKind::RefNull : TypeKind::Ref;
+        return Type { kind, type.nullable, projection->index() };
+    }
+
+    return type;
+}
+
+const TypeDefinition& TypeDefinition::unroll(TypeIndex projectee) const
+{
+    if (is<FunctionSignature>()) {
```

Comment:
Why don't we unroll struct definitions?

Reply:
We do have to unroll struct definitions, but currently there is no code that would ever call `unroll` for a struct because no struct operations have been implemented yet. So an unroll on structs would not be exercised in any tests.

How about I add a case for `StructType` here with `ASSERT_NOT_REACHED` and a FIXME comment?

Reply:
I added the assertion and fixme comment.